File: algorithms/sorting/code/cpp/selectionsort.cpp
#include <iostream>
#include <vector>

template <typename T>
void selectionSort(std::vector<T>& v)
{
	for (int i = 0; i < v.size() - 1; ++i)
	{
		T min = i;

		for (int j = i + 1; j < v.size(); ++j)
		{
			if (v[j] < v[min])
			{
				min = j;
			}
		}

		if (min != i)
		{
			T temp = v[i];
			v[i]   = v[min];
			v[min] = temp;
		}
	}
}

int main()
{
	std::vector<int> v{4, 3, 8, 6, 1, 2, 7, 5, 0, 9};

	selectionSort(v);

	for(auto const& elem : v)
	{
		std::cout << elem << ' ';
	}
	std::cout << '\n';

	return 0;
}

File: algorithms/sorting/code/js/selection-sort.js
//
// const selectionSort = (arr) => {
//
//   const swap = (arr, i, j) => {
//     let temp = arr[i];
//     arr[i] = arr[j];
//     arr[j] = temp;
//   }
//
//   for (let i = 0; i < arr.length - 1; ++i) {
//     let min = i;
//     for (j = i + 1; j < arr.length; ++j) {
//       if (arr[j] < arr[min]) {
//         min = j;
//       }
//     }
//     if (min !== i) {
//       swap(arr, i, min);
//     }
//   }
//   return arr;
// }


// EX2015 method
const selectionSort = (arr) => {

  const swap = (arr, i, j) => { [arr[i], arr[j]] = [arr[j],arr[i]]; }

  for (let i = 0; i < arr.length - 1; ++i) {
    let min = i;
    for (j = i + 1; j < arr.length; ++j) {
      if (arr[j] < arr[min]) min = j;
    }
    if (min !== i) swap(arr, i, min);
  }
  return arr;
}

console.log(selectionSort([2,6,3,7,8,4,6,7,4,2,1]));
